File: src/cpp/NindRetrolexicon/NindRetrolexicon.cpp
//
// C++ Implantation: NindRetrolexicon
//
// Description: La gestion du lexique inverse sous forme de fichier plat
// voir "nind, indexation post-S2", LAT2014.JYS.440
//
// Cette classe gehre le lexique inverse qui permet de retrouver un mot ah partir de son
// identifiant gejnejrej par le lexique.
//
// Author: jys <jy.sage@orange.fr>, (C) LATEJCON 2017
//
// Copyright: 2014-2017 LATEJCON. See LICENCE.md file that comes with this distribution
// This file is part of NIND (as "nouvelle indexation").
// NIND is free software: you can redistribute it and/or modify it under the terms of the 
// GNU Less General Public License (LGPL) as published by the Free Software Foundation, 
// (see <http://www.gnu.org/licenses/>), either version 3 of the License, or any later version.
// NIND is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Less General Public License for more details.
////////////////////////////////////////////////////////////
#include "NindRetrolexicon.h"
//#include <iostream>
using namespace latecon::nindex;
using namespace std;
////////////////////////////////////////////////////////////
// <donnejesIndexejes>     ::= <dejfinitionMot>
// <blocEnVrac>            ::= <blocUtf8>
//
// <dejfinitionMot>        ::= <motComposej> | <motSimple>
// <motComposej>           ::= <flagComposej=31> <identifiantA> <identifiantS>
// <flagComposej=31>       ::= <Integer1>
// <identifiantA>          ::= <Integer4>
// <identifiantS>          ::= <Integer4>
// <motSimple>             ::= <flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
// <flagSimple=37>         ::= <Integer1>
// <longueurMotUtf8>       ::= <Integer1>
// <adresseMotUtf8>        ::= <Integer5>
//
// <blocUtf8>              ::= { <motUtf8> }
// <motUtf8>               ::= { <Octet> }
//
// <blocEnVrac>            ::= { <Octet> }
////////////////////////////////////////////////////////////
// <spejcifique>           ::= <vide>
////////////////////////////////////////////////////////////
#define FLAG_COMPOSEJ 31
#define FLAG_SIMPLE 37
//<flagComposej=31>(1) <identifiantA>(4) <identifiantS>(4) = 9
//<flagSimple=37>(1) <longueurMotUtf8>(1) <adresseMotUtf8>(5) = 7
#define TAILLE_DEJFINITION 9
// //<motUtf8>(255) + TAILLE_IDENTIFICATION (12) = 268
#define TAILLE_DEJFINITION_MAXIMUM 268
// //taille maximum d'un mot compose (pour detecter les bouclages)
#define TAILLE_COMPOSEJ_MAXIMUM 30
//taille des spejcifiques
#define TAILLE_SPEJCIFIQUES 0
////////////////////////////////////////////////////////////
//brief Creates NindRetrolexicon.
//param fileNameExtensionLess absolute path file name without extension
//param isLexiconWriter true if lexicon writer, false if lexicon reader  
//param definitionBlocSize number of entries in a single definition block */
NindRetrolexicon::NindRetrolexicon(const string &fileNameExtensionLess,
                                   const bool isLexiconWriter,
                                   const Identification &lexiconIdentification,
                                   const unsigned int definitionBlocSize):
    NindPadFile(fileNameExtensionLess + ".nindretrolexicon", 
                isLexiconWriter, 
                lexiconIdentification, 
                TAILLE_SPEJCIFIQUES, 
                TAILLE_DEJFINITION, 
                definitionBlocSize)
{
}
////////////////////////////////////////////////////////////
NindRetrolexicon::~NindRetrolexicon()
{
}
////////////////////////////////////////////////////////////
//brief add a list of word idents in retro lexicon. If one of idents still exists, exception is raised
//param retroWords list of words definitions 
//param lexiconWordsNb number of words contained in lexicon 
//param lexiconIdentification unique identification of lexicon */
void NindRetrolexicon::addRetroWords(const list<struct RetroWord> &retroWords,
                                     const Identification &lexiconIdentification)
{
    if (!m_isWriter) 
        throw NindRetrolexiconException("NindRetrolexicon::addRetroWords retro lexicon is not writable" + m_fileName);
    //taille des spejcifiques et de l'identification qui sont en queue de buffer
    const int tailleQueue = getSpecificsAndIdentificationSize();
    //pour savoir si l'identification a ejtej ejcrite
    bool identOk = false;
    //il y a autant d'ajout au fichier qu'il y a de nouveaux identifiants
    for (list<struct RetroWord>::const_iterator it1 = retroWords.begin(); it1 != retroWords.end(); it1++) {
        const struct RetroWord &retroWord = (*it1);
        //1) rejcupehre l'adresse de la dejfinition
        unsigned long int dejfinition = getEntryPos(retroWord.identifiant);
        //2) si hors limite, ajoute un bloc de dejfinitions
        if (dejfinition == 0) addEntriesBlock(lexiconIdentification);
        //et rejcupehre l'adresse de la dejfinition
        dejfinition = getEntryPos(retroWord.identifiant);
        if (dejfinition == 0) throw NindRetrolexiconException("NindRetrolexicon::addRetroWords : " + m_fileName);
        //3) si mot simple, ejcrit d'abord l'utf8 puis la dejfinition
        if (retroWord.identifiantA == 0) {
            identOk = true;
            //se positionne ah la fin (sur les spejcifiques)
            m_file.setPos(-tailleQueue, SEEK_END);       
            const unsigned long int utf8Pos = m_file.getPos();
            m_file.createBuffer(TAILLE_DEJFINITION_MAXIMUM + tailleQueue);
            m_file.putStringAsBytes(retroWord.motSimple);
            //ajoute les spejcifiques et l'identification
            writeSpecificsHeader();
            writeIdentification(lexiconIdentification);
            m_file.writeBuffer();
            //se positionne sur la dejfinition
            m_file.setPos(dejfinition, SEEK_SET);
            m_file.createBuffer(TAILLE_DEJFINITION);
            //<flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
            m_file.putInt1(FLAG_SIMPLE);
            m_file.putInt1(retroWord.motSimple.length());
            m_file.putInt5(utf8Pos);
            m_file.writeBuffer();
        }
        //si mot composej, ejcrit la dejfinition
        else {
            //se positionne sur la dejfinition
            m_file.setPos(dejfinition, SEEK_SET);
            m_file.createBuffer(TAILLE_DEJFINITION);
            //<flagComposej=31> <identifiantA> <identifiantS>
            m_file.putInt1(FLAG_COMPOSEJ);
            m_file.putInt4(retroWord.identifiantA);
            //m_file.putInt4(retroWord.identifiantS - retroWord.identifiant);
            m_file.putInt4(retroWord.identifiantS);
            m_file.writeBuffer();
        }
    }
    //ejcrit l'identification si elle n'a pas dejjah ejtej ejcrite
    if (!identOk) {
        m_file.setPos(-tailleQueue, SEEK_END); 
        m_file.createBuffer(tailleQueue);
        writeSpecificsHeader();
        writeIdentification(lexiconIdentification);
        m_file.writeBuffer();
    }
}
////////////////////////////////////////////////////////////
//brief get word components from the specified ident
//param ident ident of word
//param components list of components of a word 
//(1 component = simple word, more components = compound word) */
//return true if word was found, false otherwise */
bool NindRetrolexicon::getComponents(const unsigned int ident,
                                     list<string> &components)
{
    //raz resultat
    components.clear();
    //lecture du mot sur le retro lexique
    struct RetroWord retroWord;
    bool existe = getRetroWord(ident, retroWord);
    //si le mot est inconnu, retour false
    if (!existe) return false;
    //le mot existe, on le decode
    while (true) {
        if (retroWord.identifiantA == 0) {
            //c'est un mot simple et c'est la fin 
            components.push_front(retroWord.motSimple);
            return true;
        }
        //c'est un mot compose
        //recupere le mot simple du couple
        struct RetroWord retroWordS;
        existe = getRetroWord(retroWord.identifiantS, retroWordS);
        if (!existe) throw NindRetrolexiconException("NindRetrolexicon::getComponents A : " + m_fileName);
        if (retroWordS.identifiantA != 0) NindRetrolexiconException("NindRetrolexicon::getComponents B : " + m_fileName);
        components.push_front(retroWordS.motSimple);
        //recupere l'autre mot du couple
        existe = getRetroWord(retroWord.identifiantA, retroWord);
        if (!existe) throw NindRetrolexiconException("NindRetrolexicon::getComponents C : " + m_fileName);  
        //pour detecter les bouclages induits par un fichier bouclant
        if (components.size() == TAILLE_COMPOSEJ_MAXIMUM) 
            throw NindRetrolexiconException("NindRetrolexicon::getComponents D : " + m_fileName);  
    }    
}    
////////////////////////////////////////////////////////////
//Recupere sur le fichier retro lexique la definition d'un mot specifie par son identifiant
//ident identifiant du mot
//retroWord structure ou est ecrite la definition
//retourne true si le mot existe, sinon false
bool NindRetrolexicon::getRetroWord(const unsigned int ident,
                                    struct RetroWord &retroWord)   
{
    const bool existe = getDejfinition(ident);
    if (!existe) return false;
    const unsigned char flag = m_file.getInt1();
    if (flag == 0) return false;
    //<flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
    else if (flag == FLAG_SIMPLE) {
        const unsigned char longueurMotUtf8 = m_file.getInt1();
        const unsigned long adresseMotUtf8 = m_file.getInt5();
        //se positionne sur la definition
        m_file.setPos(adresseMotUtf8, SEEK_SET);    
        m_file.readBuffer(longueurMotUtf8);
        const string motUtf8 = m_file.getStringAsBytes(longueurMotUtf8);
        retroWord = RetroWord(ident, motUtf8);
        return true;
    }
    //<flagComposej=31> <identifiantA> <identifiantS>
    else if (flag == FLAG_COMPOSEJ) {
        const unsigned int identifiantA = m_file.getInt4();
        //const unsigned int identifiantS = ident + m_file.getSInt4();
        const unsigned int identifiantS = m_file.getInt4();
        retroWord = RetroWord(ident, identifiantA, identifiantS);
        return true;
    }
    else throw NindRetrolexiconException("NindRetrolexicon::getRetroWord : " + m_fileName);
}
////////////////////////////////////////////////////////////
//brief Read from file datas of a specified definition and leave result into read buffer 
//param ident ident of definition
//return true if ident was found, false otherwise */
bool NindRetrolexicon::getDejfinition(const unsigned int ident)
{
    unsigned long int dejfinitionPos = getEntryPos(ident);
    if (dejfinitionPos == 0) return false;
    m_file.flush();
    //se positionne sur la definition
    m_file.setPos(dejfinitionPos, SEEK_SET);    
    m_file.readBuffer(TAILLE_DEJFINITION);
    return true;
}
////////////////////////////////////////////////////////////

File: src/py/NindRetrolexicon.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
__author__ = "jys"
__copyright__ = "Copyright (C) 2017 LATEJCON"
__license__ = "GNU LGPL"
__version__ = "2.0.1"
# Author: jys <jy.sage@orange.fr>, (C) LATEJCON 2017
# Copyright: 2014-2017 LATEJCON. See LICENCE.md file that comes with this distribution
# This file is part of NIND (as "nouvelle indexation").
# NIND is free software: you can redistribute it and/or modify it under the terms of the 
# GNU Less General Public License (LGPL) as published by the Free Software Foundation, 
# (see <http://www.gnu.org/licenses/>), either version 3 of the License, or any later version.
# NIND is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Less General Public License for more details.
import sys
from os import getenv, path
import codecs
from NindPadFile import NindPadFile
from NindPadFile import chercheVides

def usage():
    if getenv("PY") != None: script = sys.argv[0].replace(getenv("PY"), '$PY')
    else: script = sys.argv[0]
    print ("""© l'ATEJCON.
o Analyse un fichier nindretrolexicon du système nind et donne les statistiques
o Peut dumper nindretrolexicon sur <fichier>-lexique.txt et donner les
  statistiques sur les longueurs de mots. Il s'agit du lexique en clair
o Peut donner le mot spécifié par son identifiant.
Le format du fichier est défini dans le document LAT2017.JYS.470.

usage   : %s <fichier> [ <analyse> | <lexique> | <mot> <ident> ]
exemple : %s FRE.nindretrolexicon
exemple : %s FRE.nindretrolexicon lexi
exemple : %s FRE.nindretrolexicon mot 19321
"""%(script, script, script, script))

def main():
    try:
        if len(sys.argv) < 2 : raise Exception()
        retrolexiconFileName = path.abspath(sys.argv[1])
        action = 'analyse' 
        if len(sys.argv) > 2 : action = sys.argv[2]
        ident = 0
        if len(sys.argv) > 3 : ident = int(sys.argv[3])
        
        #la classe
        nindRetrolexicon = NindRetrolexicon(retrolexiconFileName)
        if action.startswith('anal'): nindRetrolexicon.analyseFichierRetrolexicon(True)
        elif action.startswith('lexi'):
            outFilename = retrolexiconFileName + '-lexique.txt'
            outFile = codecs.open(outFilename, 'w', 'utf-8')
            (nbLignes, nbErreurs, rejpartition) = nindRetrolexicon.dumpeFichier(outFile)
            outFile.close()
            rejpartition.sort()
            for (tailleMot, nombre) in rejpartition: print ('% 3d : % 9d'%(tailleMot, nombre))
            print('      % 9d lignes en erreur'%(nbErreurs))
            print('        -------')
            print('      % 9d lignes écrites dans %s'%(nbLignes, path.basename(outFilename)))
        elif action.startswith('mot'): print (nindRetrolexicon.donneMot(ident)[0])
        else: raise Exception()
    except Exception as exc:
        if len(exc.args) == 0: usage()
        else:
            print ("******************************")
            print (exc.args[0])
            print ("******************************")
            raise
        sys.exit()

#####################################################
# <donnejesIndexejes>     ::= <dejfinitionMot>
# <blocEnVrac>            ::= <blocUtf8>
#
# <dejfinitionMot>        ::= <motComposej> | <motSimple>
# <motComposej>           ::= <flagComposej=31> <identifiantA> <identifiantS>
# <flagComposej=31>       ::= <Entier1>
# <identifiantA>          ::= <Entier3>
# <identifiantS>          ::= <Entier3>
# <motSimple>             ::= <flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
# <flagSimple=37>         ::= <Entier1>
# <longueurMotUtf8>       ::= <Entier1>
# <adresseMotUtf8>        ::= <Entier5>
#
# <blocUtf8>              ::= { <Utf8> }
##############################
# <spejcifique>           ::= <vide>
#####################################################

FLAG_COMPOSEJ = 31
FLAG_SIMPLE = 37
TAILLE_COMPOSEJ_MAXIMUM = 100

class NindRetrolexicon(NindPadFile):
    def __init__(self, retrolexiconFileName):
        #si le lexique inverse n'existe pas, on ne fait rien
        if not path.isfile(retrolexiconFileName): raise Exception("%s n'existe pas"%(retrolexiconFileName))
        #on initialise la classe mehre en lecture uniquement
        NindPadFile.__init__(self, retrolexiconFileName)

    def createFile(self):
        return
               
    def donneMot (self, ident):
        mot = []
        (trouvej, motSimple, identifiantA, identifiantS) = self.__donneDefMot(ident)
        #si pas trouvej, retourne chaisne vide
        if not trouvej: return ('', 0)
        tailleMot = 0
        while True:
            #print(mot)
            tailleMot +=1
            #si c'est un mot simple, c'est la fin 
            if identifiantA == 0:
                mot.insert(0, motSimple)
                break
            #si c'est un mot composej, recupehre le mot simple du couple
            (trouvej, motSimple, identifiantA2, identifiantS2) = self.__donneDefMot(identifiantS)
            if not trouvej: raise Exception("%d pas trouvé A dans %s"%(identifiantS, self.latFileName))
            if identifiantA2 != 0: raise Exception("%d pas terminal %s"%(identifiantS, self.latFileName))
            mot.insert(0, motSimple)
            #recupere l'autre mot du couple
            (trouvej, motSimple, identifiantA, identifiantS) = self.__donneDefMot(identifiantA)
            if not trouvej: raise Exception("%d pas trouvé B dans %s"%(identifiantA, self.latFileName))
            #pour detecter les bouclages induits par un fichier bouclant
            if len(mot) == TAILLE_COMPOSEJ_MAXIMUM: 
                raise Exception("%d bouclage dans %s"%(ident, self.latFileName))
        #retourne la chaisne
        return ('_'.join(mot), tailleMot)
        #return (b'_'.join(mot), tailleMot)

    def __donneDefMot(self, ident):
        #trouve l'adresse des donnees dans le fichier
        offsetDejfinition = self.donnePositionEntreje(ident)
        #si pas trouve, le mot est inconnu
        if offsetDejfinition == 0: return (False, "", 0, 0)          #mot inconnu
        #lit la dejfinition
        self.seek(offsetDejfinition, 0)
        flag = self.litNombre1()
        if flag == 0: return (False, "", 0, 0)          #mot inconnu
        if flag == FLAG_SIMPLE:
            #<flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
            longueurMotUtf8 = self.litNombre1()
            adresseMotUtf8 = self.litNombre5()
            #lit la chaisne
            self.seek(adresseMotUtf8, 0)
            return (True, self.litChaine(longueurMotUtf8), 0, 0)
            #return (True, self.litOctets(longueurMotUtf8), 0, 0)
        if flag == FLAG_COMPOSEJ:
            #<flagComposej=31> <identifiantA> <identifiantS>
            identifiantA = self.litNombre4()
            identifiantS = self.litNombre4()
            return (True, '', identifiantA, identifiantS)
            #return (True, '', self.litNombre4(), self.litNombre4())
        raise Exception("%d mauvaise définition %s"%(ident, self.latFileName))
    
    ##################################################################
    #analyse complehtement le fichier et retourne True si ok
    def analyseFichierRetrolexicon(self, trace):
        cestbon = self.analyseFichierPadFile(trace)
        try:
            totalUtf8 = 0
            nbreSimples = nbreComposejs = 0
            #met en place la carte des non-vides
            maxIdent, nonVidesList = self.donneCarteNonVides()
            for identifiant in range(maxIdent):
                position = self.donnePositionEntreje(identifiant)
                if position == 0: 
                    raise Exception('%s : identifiant %d sans indirection '%(self.latFileName, identifiant))
                self.seek(position, 0)
                flag = self.litNombre1()
                if flag == 0: continue
                if flag == FLAG_SIMPLE:
                    #<flagSimple=37> <longueurMotUtf8> <adresseMotUtf8>
                    longueurMotUtf8 = self.litNombre1()
                    adresseMotUtf8 = self.litNombre5()
                    nonVidesList.append((adresseMotUtf8, longueurMotUtf8))
                    totalUtf8 += longueurMotUtf8
                    nbreSimples += 1
                elif flag == FLAG_COMPOSEJ:
                    #<flagComposej=31> <identifiantA> <identifiantS>
                    identifiantA = self.litNombre4()
                    identifiantS = self.litNombre4()
                    nbreComposejs += 1
        except Exception as exc: 
            cestBon = False
            if trace: print ('ERREUR :', exc.args[0])
            raise
        if trace:
            print ("=============")
            print ("%d / %d index utilisés"%(nbreSimples + nbreComposejs, maxIdent))
            print ("=============")
            
        try:
            (nbreVides, tailleVides, typesVides, typesNonVides) = chercheVides(nonVidesList)
            total = tailleVides + totalUtf8
            if trace:
                print ("CHAÎNES UTF-8  % 10d (%6.2f %%) % 9d occurrences"%(totalUtf8, float(100)*totalUtf8/total, nbreSimples))
                print ("VIDES          % 10d (%6.2f %%) % 9d occurrences"%(tailleVides, float(100)*tailleVides/total, nbreVides))
                print ("TOTAL          % 10d %08X"%(total, total))
                print ("=============")
                total = nbreSimples + nbreComposejs
                print ("MOTS SIMPLES   % 10d (%6.2f %%)"%(nbreSimples, float(100)*nbreSimples/total))
                print ("MOTS COMPOSÉS  % 10d (%6.2f %%)"%(nbreComposejs, float(100)*nbreComposejs/total))
                print ("TOTAL          % 10d"%(total))
                print ("=============")
                print ("%0.2f octets / mot"%(float(self.donneTailleFichier())/total))
                print ("=============")
        except Exception as exc: 
            cestBon = False
            if trace: print ('ERREUR :', exc.args[0])
            raise

    #######################################################################
    #dumpe le fichier lexique sur un fichier texte
    def dumpeFichier(self, outFile):
        nbLignes = nbErreurs = 0
        rejpartition = {}
        #trouve le max des identifiants
        maxIdent = self.donneMaxIdentifiant()
        for index in range(maxIdent):
            try:
                (mot, tailleMot) = self.donneMot(index)
                if tailleMot == 0: continue
                outFile.write('%06d  %s\n'%(index, mot))
                if tailleMot not in rejpartition: rejpartition[tailleMot] = 0
                rejpartition[tailleMot] +=1
            except Exception as exc: 
                outFile.write('%06d  *******ERREUR: %s\n'%(index, exc.args[0]))
                nbErreurs +=1
                raise
            nbLignes +=1
        return (nbLignes, nbErreurs, list(rejpartition.items()))
    #######################################################################


if __name__ == '__main__':
    main()
